File: src/main/java/sk/hudak/prco/dao/db/GroupEntityDao.java
package sk.hudak.prco.dao.db;

import sk.hudak.prco.api.EshopUuid;
import sk.hudak.prco.dao.BaseDao;
import sk.hudak.prco.dto.group.GroupFilterDto;
import sk.hudak.prco.model.GroupEntity;
import sk.hudak.prco.model.ProductEntity;

import java.util.List;
import java.util.Optional;

public interface GroupEntityDao extends BaseDao<GroupEntity> {

    Optional<GroupEntity> findGroupByName(String groupName);

    boolean existGroupByName(String name, Long groupIdToSkip);

    List<GroupEntity> findGroupsWithoutProduct(Long productId);

    List<GroupEntity> findGroups(GroupFilterDto groupFilterDto);

    //TODO move to ProductService...
    List<ProductEntity> findProductsInGroup(Long groupId, EshopUuid... eshopsToSkip);

    List<String> findAllGroupNames();

    /**
     * Zoznam vsetkych group, v ktorych sa dany product nachadza.
     *
     * @param productId
     * @return
     */
    List<GroupEntity> findGroupsForProduct(Long productId);
}

File: src/main/java/sk/hudak/prco/dao/db/GroupEntityDao.kt
package sk.hudak.prco.dao.db

import sk.hudak.prco.api.EshopUuid
import sk.hudak.prco.dao.BaseDao
import sk.hudak.prco.dto.GroupFilterDto
import sk.hudak.prco.model.GroupEntity
import sk.hudak.prco.model.ProductEntity

interface GroupEntityDao : BaseDao<GroupEntity> {

    //FIXME navratova hodnata nech je boolen ci existuje!!
    fun findGroupByName(groupName: String): GroupEntity?

    //TODO zmenit navratovu hodnotu na Optional<GroupEntity>
    fun existGroupByName(groupName: String, groupIdToSkip: Long?): Boolean

    fun findGroupsWithoutProduct(productId: Long?): List<GroupEntity>

    fun findGroups(groupFilterDto: GroupFilterDto): List<GroupEntity>

    //TODO move to ProductService...
    fun findProductsInGroup(groupId: Long?, withPriceOnly: Boolean, vararg eshopsToSkip: EshopUuid): List<ProductEntity>

    fun findAllGroupNames(): List<String>

    /**
     * Zoznam vsetkych group, v ktorych sa dany product nachadza.
     *
     * @param productId
     * @return
     */
    fun findGroupsForProduct(productId: Long?): List<GroupEntity>
}
